src/ext/php7/engine_hooks.c

```diff
@@ -404,17 +526,81 @@ static ddtrace_span_fci *ddtrace_prehook(zend_execute_data *call, ddtrace_dispat
     span_fci->dispatch = dispatch;
     ddtrace_open_span(span_fci);
 
-    if (dispatch->options & DDTRACE_DISPATCH_PREHOOK) {
-        continue_tracing = _dd_call_sandboxed_tracing_closure(span_fci, &dispatch->prehook, user_retval);
-        if (!continue_tracing) {
-            ddtrace_drop_top_open_span();
-            span_fci = NULL;
-        }
+    return span_fci;
+}
+
+static ddtrace_span_fci *ddtrace_fcall_begin_tracing_prehook(zend_execute_data *call, ddtrace_dispatch_t *dispatch) {
+    bool continue_tracing = true;
+    ZEND_ASSERT(dispatch);
+
+    // todo: extract common code to helper; weird to call posthook from prehook
+    ddtrace_span_fci *span_fci = ddtrace_fcall_begin_tracing_posthook(call, dispatch);
+    ZEND_ASSERT(span_fci);
+
+    zval user_retval = {.u1.type_info = IS_NULL};
+    continue_tracing = _dd_call_sandboxed_tracing_closure(span_fci, &dispatch->prehook, &user_retval);
+    if (!continue_tracing) {
+        ddtrace_drop_top_open_span();
+        span_fci = NULL;
     }
 
     return span_fci;
 }
 
+static ddtrace_span_fci *ddtrace_fcall_begin_non_tracing_posthook(zend_execute_data *call,
```

Comment:
I realized just now that we can use `dd_` for static functions if we want. My latest push went ahead with `dd_` because it's a bit shorter, but let me know if you'd prefer `_dd_` and we can have a conversation about it.

Reply:
I'm good with `dd_`. One less character 👍